### PR TITLE
SDK 2.X.X Patch 

### PR DESCRIPTION
**Pull request by emawby**
This Branch/PR is based off of version 2.16.1 version before the 3.0.0 Major Release was merged. I have cherry picked all non-merge commits except Carousel Page Impressions to this version. This branch/PR can be used to maintain a 2.X.X version to be used in wrapper SDKs

## 
This change is [Reviewable](https://reviewable.io/reviews/onesignal/onesignal-ios-sdk/855)

### CODE REVIEW

File: iOS_SDK/OneSignalSDK/Source/Requests.m [hunk 1]
```diff
@@ -143,14 +149,15 @@ + (instancetype _Nonnull)withUserId:(NSString * _Nonnull)userId appId:(NSString
...
-+ (instancetype _Nonnull)withAppId:(NSString * _Nonnull)appId withDeviceType:(NSNumber * _Nonnull)deviceType withEmail:(NSString * _Nullable)email withPlayerId:(NSString * _Nullable)playerId withEmailAuthHash:(NSString * _Nullable)emailAuthHash {
++ (instancetype _Nonnull)withAppId:(NSString * _Nonnull)appId withDeviceType:(NSNumber * _Nonnull)deviceType withEmail:(NSString * _Nullable)email withPlayerId:(NSString * _Nullable)playerId withEmailAuthHash:(NSString * _Nullable)emailAuthHash withExternalIdAuthToken:(NSString * _Nullable)externalIdAuthToken {
     let request = [OSRequestCreateDevice new];
     
     request.parameters = @{
        @"app_id" : appId,
        @"device_type" : deviceType,
        @"identifier" : email ?: [NSNull null],
        @"email_auth_hash" : emailAuthHash ?: [NSNull null],
+       @"external_user_id_auth_hash" : externalIdAuthToken ?: [NSNull null],
```
Comment (jkasten2) [inline iOS_SDK/OneSignalSDK/Source/Requests.m:160]: Not sure why this is showing up as an addition. This `external_user_id_auth_hash` line was already present in the 2.16.1 release
https://github.com/OneSignal/OneSignal-iOS-SDK/blob/2.16.1/iOS_SDK/OneSignalSDK/Source/Requests.m#L160

**Review by jkasten2 — CHANGES_REQUESTED** [review]
The changes themselves look good but seeing a lot additions that should have already been in 2.16.1. Maybe we merged in commits that we shouldn't?

File: iOS_SDK/OneSignalSDK/Source/Requests.m (hunk 1 above)
Comment (emawby) [reply iOS_SDK/OneSignalSDK/Source/Requests.m:160]: The 2.16.1 release branch was deleted and I was not able to recover it so this is based off of the 2.16.0 release branch with the 2.16.1 commits added to it. If there is a way to recover that branch I could rebase to it

Comment (jkasten2) [reply iOS_SDK/OneSignalSDK/Source/Requests.m:160]: ah, there is still a `2.16.1` tag which should point to a specific commit on master which might work. If that isn't easy todo I reviewed this PR by using the following view and it looks good to me.
https://github.com/OneSignal/OneSignal-iOS-SDK/compare/2.16.1..sdk_2.X.X

**Review by jkasten2 — APPROVED** [review]
Changes look good to me. Reviewed by using the following compare
https://github.com/OneSignal/OneSignal-iOS-SDK/compare/2.16.1..sdk_2.X.X

**Review by Jeasmine — APPROVED** [review]

**Review by adityaraute — APPROVED** [review]

**Review by adityaraute — APPROVED** [review]

**Pull request closed without merging**